samza-core/src/main/java/org/apache/samza/zk/ZkLock.java

```diff
@@ -0,0 +1,129 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.zk;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.ZkConfig;
+import org.apache.samza.coordinator.Lock;
+import org.apache.samza.coordinator.LockListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Lock primitive for Zookeeper.
+ */
+public class ZkLock implements Lock {
+
+  public static final Logger LOG = LoggerFactory.getLogger(ZkLock.class);
+  private final static String LOCK_PATH = "lock";
+  private final ZkConfig zkConfig;
+  private final ZkUtils zkUtils;
+  private final String lockPath;
+  private final String participantId;
+  private final ZkKeyBuilder keyBuilder;
+  private final Random random = new Random();
+  private String nodePath = null;
+  private LockListener zkLockListener = null;
+  private AtomicBoolean hasLock;
+
+  public ZkLock(String participantId, ZkUtils zkUtils, ZkConfig zkConfig) {
+    this.zkConfig = zkConfig;
+    this.zkUtils = zkUtils;
+    this.participantId = participantId;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    lockPath = String.format("%s/%s", keyBuilder.getRootPath(), LOCK_PATH);
+    zkUtils.makeSurePersistentPathsExists(new String[] {lockPath});
+    this.hasLock = new AtomicBoolean(false);
+  }
+
+  /**
+   * Tries to acquire a lock in order to create intermediate streams. On failure to acquire lock, it keeps trying until the lock times out.
+   * Creates a sequential ephemeral node to acquire the lock. If the path of this node has the lowest sequence number, the processor has acquired the lock.
+   */
+  @Override
+  public boolean lock() {
+    nodePath = zkUtils.getZkClient().createEphemeralSequential(lockPath + "/", participantId);
+
+    //Start timer for timeout
+    ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat("ZkLock-%d").build());
+    scheduler.schedule(() -> {
```

Comment:
As we'd discussed offline, this exception will not get propagated to the other main 
thread which is running the while loop(Since the exception thrown by the
threadPool's  thread doesn't reach the main thread's while loop).

So essentially all the followers will go into infinite loop (as long as there's a leader). 

Please share a separate state (something like CountdownLatch, AtomicBoolean, 
volatile boolean) between the two threads to indicate that the periodic 
polling should stop. 

Please correct me if i'm wrong.